### DISCUSSION TITLE
Jordan form blocks sort order

### DISCUSSION THREAD

**Discussion by rutkas** [sympy/sympy#26189] [General]
jordan_form() finds P and J matrices. where J is jordan form. It seems that the blocks are sorted from 0, lowest eigenvalues to higher eigenvalues.
Is there a way to build P and J with another sort order of blocks in J ? for example from higher eigenvalue to lower and with zeroes at the end ?

**Comment by oscarbenjamin** [comment]
You can just reorder the rows and columns:
```python
In [7]: M = Matrix([[2, 1, 0, 0], [0, 2, 0, 0], [0, 0, 3, 0], [0, 0, 4, 0]])

In [8]: P, J = M.jordan_form()

In [9]: J
Out[9]: 
⎡0  0  0  0⎤
⎢          ⎥
⎢0  2  1  0⎥
⎢          ⎥
⎢0  0  2  0⎥
⎢          ⎥
⎣0  0  0  3⎦

In [10]: ix = sorted(range(4), key=lambda i: J[i,i], reverse=True)

In [11]: J[ix,ix]
Out[11]: 
⎡3  0  0  0⎤
⎢          ⎥
⎢0  2  1  0⎥
⎢          ⎥
⎢0  0  2  0⎥
⎢          ⎥
⎣0  0  0  0⎦

```

**Reply by rutkas** [reply]
Thanks, and what about P matrix to be correspondent for the new sorted J ? It is not a reordered P

I have tried to reorder eignvalues, maybe I should build jordan form manually from them using some extra functions other then jordan_form() ?

**Reply by oscarbenjamin** [reply]
Just reorder P as well.

**Reply by oscarbenjamin** [reply]
To be clear:
```python
In [9]: P[:,ix]*J[ix,ix]*P[:,ix].inv() == M
Out[9]: True
```

**Comment by rutkas** [comment]
Or.. maybe there is a way to specify sort order to the jordan_form() function ? (...or feature idea)

**Comment by rutkas** [comment]
Thanks. I've checked on different examples.. Works like a magic :)
How does "sorted" know and analyze entire jordan blocks but not only diagonal elements ? I thought if [i,i] there then it looks only on diagonal elements ? And how does P recalculate so easy with P[:,ix] ?
Could you please describe how these algorythms work, If you have some time of course.. Very interesting
